### DISCUSSION TITLE
!pip install

### DISCUSSION THREAD

**Discussion by Niko-La** [jupyterlite/jupyterlite#297] [Q&A]
I assume since jupyterlite has no shell 
`!pip install ...`.  Does not work?

`oserror sh shell not found`

Is there a workaround ?

Love this project. Great job.

**Comment by bollwyvl** [comment]
Yep, it's just a computer with a filesystem and python. But _many_ magics
are not going to work... Or will need patches/upstream PRs.

At some point we could probably add BusyBox, xonsh, or something, likely
first as a separate kernel.

**Reply by Niko-La** [reply]
so no way to use `micropip`

i will play around and see if i can build jupyterlite with the python libraries in `requirements.txt`

**Comment by bollwyvl** [comment]
Micropip is already installed when using the pyolite kernel:

import micropip
await micropip.install(["some-package"])

...Will work for any package what...
- was already built for pyodide 0.17, e.g. numpy
- are pure python wheels on PyPI

(And all dependencies)

Please have a look at the documentation site (jupyterlite.rtfd.io) and the
pyolite examples in the demo to get a feel for what is already available.

Building the core client web application from source is considerably more
involved, and requires nodejs and 1gb of npm downloads. Contributions
welcome, of course! But building from source doesn't really have a lot of
benefits right now, as there aren't that many knobs to turn vs the CLI
pre-release package available on pypi... And some rather large changes are
coming (pyodide 0.18, python 3.9)

**Reply by Niko-La** [reply]
Is pyodide 0.18 /py 3.1p running on jupyterlite yet? Any update to pip install. Thx for update.

**Reply by jtpio** [reply]
Currently it's Pyodide 0.18.1 with Python 3.9.5:

![image](https://user-images.githubusercontent.com/591645/141448591-b3e559c0-da47-4efc-8190-2e6f67f69f8e.png)

**Comment by dsblank** [comment]
See also the idea on https://github.com/jupyterlite/jupyterlite/issues/237 to replace `%pip` with a version that uses micropip.
